tests/client_test.py

```diff
@@ -610,3 +630,317 @@ def test_intercept_unary_unary_failed_request(self):
+
+    def test_get_request_id(self):
+        mock_metadata = (('request-id', '123456'),)
+        interceptor = self._create_test_interceptor()
+        result = interceptor._get_request_id(mock_metadata)
+        self.assertEqual(result, '123456')
+
+    def test_get_request_id_no_id(self):
+        mock_metadata = (('another-key', 'another-val'),)
+        interceptor = self._create_test_interceptor()
+        result = interceptor._get_request_id(mock_metadata)
+        self.assertEqual(result, None)
+
+    def test_get_google_ads_failure(self):
+        interceptor = self._create_test_interceptor()
+        mock_metadata = ((interceptor._FAILURE_KEY, self._MOCK_FAILURE_VALUE),)
+        result = interceptor._get_google_ads_failure(mock_metadata)
+        self.assertIsInstance(result, error_protos.GoogleAdsFailure)
+
+    def test_get_google_ads_failure_decode_error(self):
+        interceptor = self._create_test_interceptor()
+        mock_failure_value = self._MOCK_FAILURE_VALUE + b'1234'
+        mock_metadata = ((interceptor._FAILURE_KEY, mock_failure_value),)
+        result = interceptor._get_google_ads_failure(mock_metadata)
+        self.assertEqual(result, None)
+
+    def test_get_google_ads_failure_no_failure_key(self):
+        mock_metadata = (('another-key', 'another-val'),)
+        interceptor = self._create_test_interceptor()
+        result = interceptor._get_google_ads_failure(mock_metadata)
+        self.assertEqual(result, None)
+
+    def test_get_google_ads_failure_with_None(self):
+        interceptor = self._create_test_interceptor()
+        result = interceptor._get_google_ads_failure(None)
+        self.assertEqual(result, None)
+
+    def test_handle_grpc_exception(self):
+        """If exception is not retryable and is a GoogleAdsFailure convert to
+           GoogleAdsException and raise.
+        """
+        mock_error_message = self._MOCK_FAILURE_VALUE
+
+        class MockRpcError(grpc.RpcError):
+            def code(self):
+                return grpc.StatusCode.INVALID_ARGUMENT
+
+            def trailing_metadata(self):
+                return ((interceptor._FAILURE_KEY, mock_error_message),)
+
+        interceptor = self._create_test_interceptor()
+
+        self.assertRaises(
+            GoogleAdsException,
+            interceptor._handle_grpc_exception,
+            MockRpcError()
+        )
+
+    def test_handle_grpc_exception_retryable(self):
+        """If exception is retryable raise as-is.
+        """
+        class MockRpcError(grpc.RpcError):
+            def code(self):
+                return grpc.StatusCode.INTERNAL
+
+        interceptor = self._create_test_interceptor()
+
+        self.assertRaises(
+            MockRpcError,
+            interceptor._handle_grpc_exception,
+            MockRpcError()
+        )
+
+    def test_handle_grpc_exception_not_google_ads_failure(self):
+        """If exception is not retryable and not a GoogleAdsFailure raise as-is.
+        """
```

Comment:
If the triple-quotes would fit, should go on previous line.

Reply:
In this case the triple-quotes brings it just over 80 lines